texar/modules/decoders/rnn_decoder_base.py

```diff
@@ -45,13 +46,17 @@ class RNNDecoderBase(DecoderBase[State, Output]):
     def __init__(self,
                  input_size: int,
                  vocab_size: int,
+                 embedder: EmbedderBase,
```

Comment:
Yes, but the embedder takes only the token IDs as argument. To supply a custom embedding function, one can just subclass the decoder and override the `embed_tokens` method.

Reply:
Passing in a module and subclassing are two different means. It seems weird to mix them. It'd be ideal if each means alone can fully support the same level of customizability.

E.g., we can probably explicitly have a constructor argument like `embedder_w_time` that takes two arguments. Users can specify either `embedder` or `embedder_w_time`.

Reply:
I agree that these are two different ways of doing things, but I think the more common use case for RNNs is just a vanilla embedder without positions. On the other hand, Transformers need position embeddings, so for `TransformerDecoder` its constructor takes `embedding_fn` instead of `embedder`. For less common use cases, we just ask the user to subclass and override.

Reply:
I felt subclassing is not that natural in Texar cuz all other Texar modules do no apply this way. Most customization is done by passing customized modules (e.g., Cells for RNNs) -- that's why it took me a while to consider this when we first discussed :). It's perhaps not necessary to introduce this way only for decoder's embedder --- Of course, users can do subclassing, but this shouldn't be a "must" way even for less common cases.

I feel it's not necessary to make RNN and Transformer Decoders different. Just use `embedding_fn` for RNNDecoders too (assuming `embedding_fn` takes one argument) --- Or a more clear way could be to have two arguments, one for single-arg embedder, the other for two-arg embedder, and again, same for both RNN and Transformer. (Or an `embedder` arg + an int arg indicating how many arguments for the `embedder`. RNN defaults to 1, and Transformer defaults to 2).

What do you think

Reply:
I see your point, and I agree with you.

Then, I would suggest changing the constructor so it takes two arguments `embedder` and `positional_embedder` (or some other name to avoid confusion, but probably not `embedding_fn`). Both `embedder` and `positional_embedder` can be PyTorch modules or functions, `embedder` takes one argument and `positional_embedder` takes two. 

Users would have three ways of specifying the embedder:
1. Provide `embedder`, and set `positional_embedder` to `None`.
2. Provide `positional_embedder`, and set `embedder` to `None`.
3. Set both `embedder` and `positional_embedder` to `None`, and override the `embed_tokens` method.
All other combinations would result in exceptions being raised. We can also issue a warning if `embedder` but not `positional_embedder` is used for Transformer models.

Reply:
This sounds good to me, though we may rename the two arguments to avoid confusion (esp., b/w `positional_embedder` which takes two arguments and `PositionEmbedder` which is for position embedding only). We can probably use `word_embedder` & `word_time_embedder`, or sth more concise.

Reply:
How about `token_embedder` and `token_position_embedder`, which matches the argument names in `embed_tokens`? I would prefer `token` to `word` because a lot of the pre-trained decoders are operating on token-level (subwords) rather than word-level. If `token_position_embedder` still sounds ambiguous, we can also go for `token_time_embedder` and correspondingly change the `positions` argument in `embed_tokens` to `times`.